[TST] Add tests for tedana.model.monoexponential

tsalo opened the pull request:
<!---
This is a suggested pull request template for tedana.
It's designed to capture information we've found to be useful
in reviewing pull requests.

If there is other information that would be helpful to include,
please don't hesitate to add it!
-->

Changes proposed in this pull request:

- Add test for tedana.model.monoexponential.fit_decay that compares results to results used for integration test of full workflow.
- Add smoke test for tedana.model.monoexponential.fit_decay_ts that just runs the function and checks that the resulting numpy arrays are in the right shape.

tsalo commented:
The Travis coverage tests cannot find `/home/neuro/data/zcat_ffd.nii.gz` in `test_model_exponential.nii.gz`, but I'm not sure why.

emdupre commented:
The travis environment doesn't have that file, only the circle ! We're not yet shipping with a downsampled `zcat` file&mdash; we could fake one by concatenating all of the downsampled echos, or you could use / downsample the averaged `zcat` volume included in circle.

emdupre commented:
Right now we're having [travis run the linting and unit tests](https://github.com/ME-ICA/tedana/blob/master/.travis.yml#L34), and [circle run the integration test](https://github.com/ME-ICA/tedana/blob/master/circle.yml#L40) (i.e., [`test_tedana`](https://github.com/ME-ICA/tedana/blob/master/tedana/tests/test_tedana.py)). 

The idea was that unit tests are usually quick, and our integration test clocks around half an hour. My worry is that we'll overload the circle testing environment and end up with two hours of testing there and only linting on travis, which will increase the overall testing time than if we share resources a little more effectively.

Basically, this is totally a judgement call ! We're early enough along in testing that I don't think we're going to get near two hours either way, so the current situation might be over optimizing. We can also use the cache on circle a little more effectively than we are right now (see #59). WDYT, @tsalo ?

tsalo commented:
That makes sense and I'm happy to follow the current convention. Still, we are going to need data for unit tests as well, right?

emdupre commented:
Absolutely ! We uploaded some downsampled [individual echo data](https://github.com/ME-ICA/tedana/tree/master/tedana/tests/data) so it could be shipped with the tests. I'd probably work off the [`zcat` file](https://github.com/ME-ICA/tedana/blob/master/circle.yml#L18) and add something new to the data folder for these tests&mdash; with something like

```
import numpy as np
from nilearn import image

file = 'zcat_ffd.nii.gz'
target_aff = np.eye(4) * 4
image.mean_img(file, target_aff).to_filename("avg_zcat_ffd.nii.gz")
```

Or, if you want to provide some sort of timeseries, we could use `image.index_img` instead ! WDYT ?

tsalo commented:
Everything is passing now, so I think it's ready for review.

rmarkello commented on tedana/tests/test_model_monoexponential.py:
```diff
@@ -0,0 +1,43 @@
+    in_files = [op.join(get_test_data_path(), 'echo{0}.nii.gz'.format(i+1))
+                for i in range(3)]
+    data, _ = utils.load_data(in_files, n_echos=len(tes))
+    mask, mask_sum = utils.make_adaptive_mask(data, minimum=False, getsum=True)
```
Is there a way to refactor this to avoid duplicating the same 5 lines (`tes = ...` through `mask, mask_sum = ...`)? Options include making them globals or creating a class structure with those lines in the `__init__` and having `test_fit_decay` and `test_fit_decay_ts` as methods.

Indeed, this goes back to #50... Having to repeat these steps for testing everything seems wasteful / redundant.

emdupre replied:
Oof. This might be outside the scope of this PR, since #50 is such an extensive issue. 

Personally, I'd rather repetition than re-introducing globals at this point&mdash; I know that their usage is less problematic here, but we just went through so much trouble to get rid of them ! It's mostly my being dogmatic, but still.

@tsalo, if you can see a clean way to cut down on the repetition, that would be great. If you want to postpone tackling this until we modularize the functions (much) better, I'm fine with that, too.

tsalo replied:
Good catch @rmarkello. I'll just add a test-specific pytest fixture to load data.

rmarkello replied:
That's a great solution!!!

rmarkello commented on tedana/tests/test_model_monoexponential.py:
```diff
@@ -0,0 +1,43 @@
+    assert len(t2ss.shape) == 2
+    assert len(s0vs.shape) == 2
+    assert len(t2svG.shape) == 1
+    assert len(s0vG.shape) == 1
```
This is partially preference, but I think `assert data.ndim == X` is a bit more parsimonious than `assert len(data.shape) == X`. I'm open to disagreement here, though! @emdupre do you have thoughts?

emdupre replied:
I think you're probably right, @rmarkello, that `.ndim` is more elegant (and, importantly, I think equally easy to read !). Would you mind updating, @tsalo ?

emdupre commented on tedana/model/monoexponential.py:
```diff
@@ -22,22 +22,22 @@ def fit_decay(data, tes, mask, masksum, start_echo):
+    s0va : (S) :obj:`numpy.ndarray`
         Limited S0 map
-    t2ss : (S x E) :obj:`numpy.ndarray`
+    t2ss : (S, ?) :obj:`numpy.ndarray`
```
Is it unclear what the returned dimensions are, here ?

rmarkello replied:
Yes, I'd noticed that too when going through the code a while back. You'll notice that actually overwrites itself; since it's `range(start_echo, n_echos + 1)` when it gets to n_echos it will index the same location. Completely bonkers and I just didn't have the headspace to figure out *why* it was happening...

emdupre replied:
You're right, and I think that's incorrect behavior. Going back to the [original code](https://github.com/ME-ICA/me-ica/blob/master/meica.libs/t2smap.py#L235), zero should be the first value. Nice catch !

rmarkello replied:
I think the `start_echo` kwarg was added in the first place to address this discrepancy (but you did that, @emdupre, so you would know better than any of us!). That being said, it might be worth digging into *why* they were ever different. For now we should retain this nonsense functionality so that `tedana.workflows.tedana()` is the same. Indeed, `tedana.workflows.t2smap()` [handles the discrepancy correctly](https://github.com/ME-ICA/tedana/blob/master/tedana/workflows/t2smap.py#L48).

emdupre replied:
Yes, I'm remembering (slowly) that I added the kwarg there to try and align them. @rmarkello, are you saying we should keep the bug ? I guess this is something we could address in #25 ... we should make a note on that issue, if so.

rmarkello replied:
No, no, we shouldn't keep it, you're right. But I do think it should potentially be addressed in #25 / a separate PR, just to retain some semblance of order as we try and make PRs more feature-based and less gigantic overhauls of the code.

To @KirstieJane: thank you! Your continued support for all things `tedana` is :100: :blush: :+1:

emdupre replied:
I think that's a reasonable course of action, @rmarkello ! @tsalo, would you mind adding that issue to #25 so we maintain a record ?

Thanks everyone 💖

emdupre commented:
Rather than dropping the `rcond` argument, could we instead force the travis environment to use an up-to-date numpy ? Maybe [here](https://github.com/ME-ICA/tedana/blob/master/.travis.yml#L18) with `pip install --upgrade requirements.txt`

emdupre commented:
We've been testing on `numpy >= 1.14`. It looks like that's the minimum, since otherwise we hit this error !

tsalo commented:
Sorry for the delay, but I've updated #25 with information about the discrepancy between the two functions. Please feel free to add to it if I got anything confused.

I've also added the minimum numpy version to the requirements file and have re-added the `rcond` argument to the `lstsq` call, so I think that this PR is ready for review again.

emdupre commented:
This LGTM ! Thanks so much for updating it, @tsalo.

@rmarkello, any other comments ?

rmarkello commented:
Looks great! I'm excited to see some coverage increases :smile: :chart_with_upwards_trend:
